File: db-async-common/src/test/scala/com/github/mauricio/async/db/pool/TimeoutSchedulerSpec.scala
/*
 * Copyright 2013 Maurício Linhares
 *
 * Maurício Linhares licenses this file to you under the Apache License,
 * version 2.0 (the "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at:
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 */
package com.github.mauricio.async.db.pool

import java.util.concurrent.{ScheduledFuture, TimeoutException}
import com.github.mauricio.async.db.util.{ByteBufferUtils, ExecutorServiceUtils}
import org.specs2.mutable.SpecificationWithJUnit
import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.duration._
import scala.concurrent.{Future, Promise}

/**
 * Tests for TimeoutScheduler
 */
class TimeoutSchedulerSpec extends SpecificationWithJUnit  {

  val TIMEOUT_DID_NOT_PASS = "timeout did not pass"

  "test timeout did not pass" in {
    val timeoutScheduler = new DummyTimeoutScheduler()
    val promise = Promise[String]()
    val scheduledFuture  = timeoutScheduler.addTimeout(promise,Some(Duration(1000, MILLISECONDS)))
    Thread.sleep(100);
    promise.isCompleted === false
    promise.success(TIMEOUT_DID_NOT_PASS)
    Thread.sleep(1500)
    promise.future.value.get.get === TIMEOUT_DID_NOT_PASS
    scheduledFuture.get.isCancelled === true
    timeoutScheduler.timeoutCount === 0
  }

  "test timeout passed" in {
    val timeoutMillis = 100
    val promise = Promise[String]()
    val timeoutScheduler = new DummyTimeoutScheduler()
    val scheduledFuture  = timeoutScheduler.addTimeout(promise,Some(Duration(timeoutMillis, MILLISECONDS)))
    Thread.sleep(1000)
    promise.isCompleted === true
    scheduledFuture.get.isCancelled === false
    promise.trySuccess(TIMEOUT_DID_NOT_PASS)
    timeoutScheduler.timeoutCount === 1
    promise.future.value.get.get must throwA[TimeoutException](message = s"Operation is timeouted after it took too long to return \\(${timeoutMillis} milliseconds\\)")
  }

  "test no timeout" in {
    val timeoutScheduler = new DummyTimeoutScheduler()
    val promise = Promise[String]()
    val scheduledFuture  = timeoutScheduler.addTimeout(promise,None)
    Thread.sleep(1000)
    scheduledFuture === None
    promise.isCompleted === false
    promise.success(TIMEOUT_DID_NOT_PASS)
    promise.future.value.get.get === TIMEOUT_DID_NOT_PASS
    timeoutScheduler.timeoutCount === 0
  }
}

File: db-async-common/src/test/java/com/github/jasync/sql/db/pool/TimeoutSchedulerSpec.kt
package com.github.jasync.sql.db.pool

import com.github.jasync.sql.db.util.ExecutorServiceUtils
import com.github.jasync.sql.db.util.NettyUtils
import com.github.jasync.sql.db.util.isCompleted
import com.github.jasync.sql.db.util.success
import org.junit.Assert
import org.junit.Test
import java.time.Duration
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ExecutionException
import java.util.concurrent.TimeoutException
import kotlin.test.assertEquals
import kotlin.test.assertFalse
import kotlin.test.assertNotNull
import kotlin.test.assertNull
import kotlin.test.assertTrue

/**
 * Tests for TimeoutScheduler
 */
class TimeoutSchedulerSpec {

    val TIMEOUT_DID_NOT_PASS = "timeout did not pass"

    private fun createTimeoutScheduler() =
        TimeoutSchedulerImpl(ExecutorServiceUtils.CommonPool, NettyUtils.DefaultEventLoopGroup, {})

    @Test
    fun `test timeout did not pass`() {
        val timeoutScheduler = createTimeoutScheduler()
        val promise = CompletableFuture<String>()
        val scheduledFuture = timeoutScheduler.addTimeout(promise, Duration.ofMillis(1000), "connectionId")
        Thread.sleep(100)
        assertFalse(promise.isCompleted)
        promise.success(TIMEOUT_DID_NOT_PASS)
        Thread.sleep(1500)
        assertEquals(TIMEOUT_DID_NOT_PASS, promise.get())
        assertTrue(scheduledFuture!!.isCancelled)
        assertEquals(false, timeoutScheduler.isTimeout())
    }

    @Test()
    fun `test timeout passed`() {
        val timeoutMillis: Long = 100
        val promise = CompletableFuture<String>()
        val timeoutScheduler = createTimeoutScheduler()
        val scheduledFuture = timeoutScheduler.addTimeout(promise, Duration.ofMillis(timeoutMillis), "connectionId")
        Thread.sleep(1000)
        assertTrue(promise.isCompleted)
        assertFalse(scheduledFuture!!.isCancelled)
        assertEquals(true, timeoutScheduler.isTimeout())
        try {
            promise.get()
            Assert.fail()
        } catch (e: ExecutionException) {
            assertNotNull(e.cause)
            assertTrue(e.cause is TimeoutException)
        }
    }

    @Test()
    fun `test timeout passed and completed after timeout`() {
        val timeoutMillis: Long = 100
        val promise = CompletableFuture<String>()
        val timeoutScheduler = createTimeoutScheduler()
        val scheduledFuture = timeoutScheduler.addTimeout(promise, Duration.ofMillis(timeoutMillis), "connectionId")
        Thread.sleep(1000)
        promise.success(TIMEOUT_DID_NOT_PASS)
        assertTrue(promise.isCompleted)
        assertFalse(scheduledFuture!!.isCancelled)
        assertEquals(true, timeoutScheduler.isTimeout())
        try {
            promise.get()
            Assert.fail()
        } catch (e: ExecutionException) {
            assertNotNull(e.cause)
            assertTrue(e.cause is TimeoutException)
        }
    }

    @Test
    fun `test no timeout`() {
        val timeoutScheduler = createTimeoutScheduler()
        val promise = CompletableFuture<String>()
        val scheduledFuture = timeoutScheduler.addTimeout(promise, null, "connectionId")
        Thread.sleep(1000)

        assertNull(scheduledFuture)
        assertFalse(promise.isCompleted)
        promise.success(TIMEOUT_DID_NOT_PASS)
        assertEquals(TIMEOUT_DID_NOT_PASS, promise.get())
        assertEquals(false, timeoutScheduler.isTimeout())
    }
}
